### ISSUE TITLE
Moya says that response is nil after coming back to the app from lock screen 

### ISSUE THREAD

**Issue by nickolayl** [Moya/Moya#494]
Hi,

Here is the case:
- return to the app from the lock screen;
- nearly immediately (1-2 seconds) send the HTTP request via moya;

Actual Result:
- moya returns error `Moya.Error.Underlying` and says that the response (NSHTTPURLResponse) is `nil`. NSError contains - `error NSError?    domain: "NSURLErrorDomain" - code: 18446744073709550611 0x000000012c5d9700`

Note:
The code where this debug info caught:
File: `Moya.swift`
Func: `internal func convertResponseToResult(response: NSHTTPURLResponse?, data: NSData?, error: NSError?) -> Result<Moya.Response, Moya.Error>`

It could be steadily reproduced **only within 1-2 seconds after returning to the application from the lock screen**.

Please help.

**Comment by ashfurrow** [comment]
Very strange, thanks for reporting this. Could you provide some more details so we can reproduce this? Is it on a device or just the simulator? iOS 8, 9, or both? The networking code that demonstrates the problem would be helpful, if you can share it. Thanks again!

**Comment by nickolayl** [comment]
Yes. Sure. We will upload our app in a private repo and give you access as we do think that the code should be seen as a whole and not as a except from the context.
In short we have chain of two requests (rxswift). The first one goes to HTTP and works well, the subsequent one calls youtube API and fails with the symptoms I described above.
It is reproduced on iOS 9 only. iOS 8 seems to not having this issue.
For App Transport Security the following settings are applied:
`<key>NSAppTransportSecurity</key>
<dict>
  <key>NSAllowsArbitraryLoads</key>
      <true/>
</dict>`

**Comment by nickolayl** [comment]
Hey Ash. Just sent you the code that reproduces the issue. We just broke our heads searching for the cause of the issue but still no luck.

**Comment by ashfurrow** [comment]
Received the email, taking a look ASAP, probably won't be until the weekend. Will update.

**Comment by ashfurrow** [comment]
Okay, I've tried taking a look but I don't have a device at the moment to test on. I'll try to take a look this weekend. In the mean time, I'd recommend trying to following trying to call the Alamofire request manually (if you can). I'd like to narrow down if this is a problem with Moya, Alamofire, or NSURLConnection. Can you provide any more details, like what device you're using, wifi status, any server logs you have? Thanks!

**Comment by nickolayl** [comment]
1. Device: any iPhone physical (not simulator) device
2. wifi connection - excellent. Internet connection 10Mbps (i.e. good)
3. Server logs. As you could see there are two requests sent in a row. The first one is to our server and it's 200OK. The next one HTTPS sent to youtube. We cannot get youtube server logs.
4. **The important thing - if you will allow ~5+ seconds to pass after the moment you unlocked device and until the moment you pull to refresh for the news - everything will work ok. Also the second pull-to-refresh works ok.**

**Comment by kevin5396** [comment]
Same problem with my App.
Turns out that if the response is empty, .mapJson() will cause this error.

**Comment by ashfurrow** [comment]
@kevin5396 Interesting! Makes sense, since `NSJSONSerialization` would fail with empty data. I'm not sure what the best way to solve this is, maybe extend `mapJSON` to include a parameter for allowing empty data?

**Comment by kevin5396** [comment]
@ashfurrow Maybe a silence option to prevent the error from throwing out but rather just return a `nil` ?

**Comment by ashfurrow** [comment]
Makes sense! We should provide a default option, and this'll need to be added to both the RxSwift and ReactiveCocoa extensions too. Anyone available to submit a pull request?

**Comment by ashfurrow** [comment]
So it looks like a PR to fix this issue would modify [the `mapJSON` function](https://github.com/Moya/Moya/blob/645398713e41a0f61a36c57e707d5135f2fa8477/Source/Response.swift#L60-L66).

```swift
func mapJSON() throws -> AnyObject {
    do {
        return try NSJSONSerialization.JSONObjectWithData(data, options: .AllowFragments)
    } catch {
        throw Error.Underlying(error as NSError)
    }
}
```

We would need to add a new parameter like `failsOnEmptyResponse: Bool = true`. If it's `false` and `data == nil`, we need to return some other value. An empty `NSArray` or `NSDictionary` might work, or maybe `NSNull()`. Not sure, open for discussion. 

Oh, and we should test it, too 😉

**Comment by nickolayl** [comment]
Ash, sorry what's the "PR"? Possible resolution?

**Comment by sunshinejr** [comment]
@nickolayl Pull Request 😉

**Comment by nickolayl** [comment]
@ashfurrow please re-open the issue. This issue still exists in the latest version of Moya.

Asked on StackOverflow
http://stackoverflow.com/questions/38900522/ios-https-get-empty-response-from-server-after-the-app-returns-back-from-the-l

**Comment by ashfurrow** [comment]
Okay, re-opening for discussion. Does Alamofire have any similar issues?

**Comment by nickolayl** [comment]
Don't know. Need to check. Who can check fast?

**Comment by ashfurrow** [comment]
The issues list is at https://github.com/Alamofire/Alamofire/issues I'm unavailable to help right now, try searching for the symptoms and see what others have experienced.

**Comment by BasThomas** [comment]
Had a quick look, but can't seem to find anything similar in the Alamofire issues.

**Comment by scottrhoyt** [comment]
Since there hasn't been any activity on this thread for some time, I am going to go ahead and close it. Please re-open or create a new issue if you would like to discuss further. Thanks.

**Comment by EvsenevDev** [comment]
@ashfurrow @scottrhoyt
Hey, could you reopen the issue?

Steps to reproduce:
1) Start upload file - **task is .uplaod(url)**
2) Press "Home" button
3) Wait 2-3 seconds
4) Open app

It should return error `received empty network response`

**Comment by sunshinejr** [comment]
Hey @EvsenevDev, thanks for steps to reproduce! Anyone from @Moya/contributors could try to reproduce the issue?

**Comment by EvsenevDev** [comment]
@sunshinejr looks like iOS cancel uploading in background and Alamofire not supports background tasks for now, but not sure.
